lib/std/Thread.zig

```diff
@@ -76,6 +76,121 @@ pub fn spinLoopHint() void {
     }
 }
 
+const maxNameLen = switch (std.Target.current.os.tag) {
+    .linux => 16,
+    .windows => 32,
+    .macos, .ios, .watchos, .tvos => 64,
+    .netbsd => 32,
+    .freebsd => 16,
+    .openbsd => 32,
+    else => 0,
+};
+
+pub fn setName(self: Thread, name: [:0]const u8) !void {
+    if (name.len >= maxNameLen) return error.NameTooLong;
+
+    switch (std.Target.current.os.tag) {
+        .linux => if (use_pthreads) {
+            const err = c.pthread_setname_np(self.data.handle, name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        } else if (self.data.handle == getCurrentId()) {
+            _ = try os.prctl(.SET_NAME, .{@ptrToInt(name.ptr)});
+        } else if (!use_pthreads) {
+            var buf: [32]u8 = undefined;
+            const path = try std.fmt.bufPrint(&buf, "/proc/self/task/{d}/comm", .{self.data.handle});
+
+            const file = try std.fs.cwd().openFile(path, .{ .write = true });
+            defer file.close();
+
+            try file.writer().print("{s}", .{name});
+        },
+        .windows => {
+            // TODO
+        },
+        .macos, .ios, .watchos, .tvos => if (use_pthreads) {
+            // There doesn't seem to be a way to set the name for an arbitrary thread, only the current one.
+            if (self.data.handle != getCurrentId()) return;
+
+            const err = c.pthread_setname_np(name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        .netbsd => if (use_pthreads) {
+            const err = c.pthread_setname_np(self.data.handle, name.ptr, null);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        .freebsd, .openbsd => if (use_pthreads) {
+            // Use pthread_set_name_np for FreeBSD because the alias pthread_setname_np is FreeBSD 12.2+ only.
+            const err = c.pthread_set_name_np(self.data.handle, name.ptr);
+            return switch (err) {
+                0 => {},
+                else => unreachable,
+            };
+        },
+        else => {},
+    }
+}
+
+pub fn getName(self: Thread, allocator: *std.mem.Allocator) !?[:0]const u8 {
+    // musl doesn't provide pthread_getname_np and there's no way to retrieve the thread id of an arbitrary thread.
+    if (use_pthreads and self.data.handle != getCurrentId() and comptime std.Target.current.abi.isMusl()) {
+        return null;
+    }
+
+    var buffer = try allocator.allocWithOptions(u8, maxNameLen - 1, null, 0);
```

Comment:
What if `maxNameLen` is zero?
Wouldn't it make more sense to have `maxNameLen` exclude the tail zero?

Reply:
My thinking was it would be less surprising for a reader to see the maxNameLen with the same value as in the different man pages, but I'm not particularly attached to that. I'll change it.

Reply:
From the Zig user POV the interesting part is the number of characters one can use, the presence of the trailing zero is an implementation detail that leaks to the user.
Moreover you get an extra byte when allocating a buffer using `var buf: [maxNameLen:0]`, dropping the trailing zero is IMO justified.

One more idea, what about a pure Zig API where the user passes in a `[]const u8` slice? Zero-terminating the string on the fly here is pretty cheap and would allow for a more idiomatic API.

Reply:
(`maxNameLen` changed in the latest commit).

Not sure I get your last point, do you mean something like `fn getName(self: Thread, buffer: []u8) []const u8` ?

Reply:
Not quite, I was thinking of something like
```zig
fn getName(self: Thread, allocator: *std.mem.Allocator) !?[]const u8;
fn setName(self: Thread, name: []const u8) !void;
```

where we get rid of the sentinel.

Reply:
Ah right. Yeah I can do that.

Reply:
Keep in mind that Windows `SetThreadDescription` requires a UTF16-LE encoded string so we may need an allocator for `setName` or set an arbitrary upper limit and allocate a static buffer.